### PR TITLE
Rework

### PR DESCRIPTION
**Pull request by riamon-v**
See #3

### CODE REVIEW

File: include/Listeners.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+namespace Listeners {
+
+  #define SET_LISTENER(TYPE, SUBTYPE, NAME, TARGET)			\
+    NAME.notify = [](wl_listener *listener, void *data)			\
+  		{							\
+  		  TYPE *that(static_cast<TYPE *>(wl_container_of(listener, static_cast<SUBTYPE *>(nullptr), NAME))); \
+  		  that->TARGET(listener, data);				\
+  		};							\
+
+    struct ServerCursorListeners
```
Comment (Gaspard--) [inline include/Listeners.hpp]: Why put all the struts outside of the corresponding classes' files ?

File: include/Output.hpp [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
   struct wl_list link;
+
+  void setFrameListener();
+
+private:
   Server *server;
   struct wlr_output *wlr_output;
-  struct wl_listener frame;
+
+private:
```
Comment (Gaspard--) [inline include/Output.hpp:27]: Why is private present twice?

File: src/Keyboard.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+void Keyboard::keyboard_handle_key(struct wl_listener *listener, void *data)
+{
+  // Keyboard *keyboard = wl_container_of(listener, keyboard, key);
+  // Server *server = keyboard->server;
+  struct wlr_event_keyboard_key *event = static_cast<struct wlr_event_keyboard_key *>(data);
+  struct wlr_seat *seat = server->seat->getSeat();
+
+  uint32_t keycode = event->keycode + 8;
+  const xkb_keysym_t *syms;
```
Comment (Gaspard--) [inline src/Keyboard.cpp:49]: `syms` looks allocated, if that's the case, where is it unallocated?

File: src/Seat.cpp [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
+{
+  struct wlr_seat_pointer_request_set_cursor_event *event = static_cast<struct wlr_seat_pointer_request_set_cursor_event *>(data);
+  struct wlr_seat_client *focused_client = seat->pointer_state.focused_client;
+  if (focused_client == event->seat_client)
+    {
+wlr_cursor_set_surface(server->cursor->cursor, event->surface, event->hotspot_x, event->hotspot_y);
+    }
+}
+
+struct wlr_seat *Seat::getSeat() const
```
Comment (Gaspard--) [inline src/Seat.cpp]:
```suggestion
struct wlr_seat *Seat::getSeat() const noexcept
```

File: src/XdgShell.cpp [hunk 5]
```diff
@@ -1,67 +1,50 @@
...
-  void xdg_toplevel_request_resize(struct wl_listener *listener, void *data)
-  {
-    struct wlr_xdg_toplevel_resize_event *event = static_cast<struct wlr_xdg_toplevel_resize_event *>(data);
-    View *view = wl_container_of(listener, view, request_resize);
-    ServerView::begin_interactive(view, CursorMode::CURSOR_RESIZE, event->edges);
-  };
+void XdgShell::xdg_surface_destroy(struct wl_listener *listener, void *data)
+{
+  wl_list_remove(&view->link);
+  //Maybe delete the view here ?
```
Comment (Gaspard--) [inline src/XdgShell.cpp]:
```suggestion
  // TODO: Maybe delete the view here ?
```
Also please open an issue or put a note somewhere so we don't forget this.

File: include/Seat.hpp [hunk 6]
```diff
@@ -1,8 +1,20 @@
...
+class Server;
+
+class Seat : public Listeners::SeatListeners
 {
+public:
+  Seat(Server *server);
+  ~Seat() = default;
+
   void seat_request_cursor(struct wl_listener *listener, void *data);
+  struct wlr_seat *getSeat() const;
```
Comment (Gaspard--) [inline include/Seat.hpp]:
```suggestion
  struct wlr_seat *getSeat() const noexcept;
```

File: include/ServerOutput.hpp [hunk 7]
```diff
@@ -18,3 +13,24 @@ struct render_data
...
+{
+public:
+  ServerOutput(Server *server);
+  ~ServerOutput() = default;
+
+  void output_frame(struct wl_listener *listener, void *data);
+  void server_new_output(struct wl_listener *listener, void *data);
+  static void render_surface(struct wlr_surface *surface, int sx, int sy, void *data);
+
+  struct wlr_output_layout *getLayout() const;
```
Comment (Gaspard--) [inline include/ServerOutput.hpp]:
```suggestion
  struct wlr_output_layout *getLayout() const noexcept;
```

File: src/ServerOutput.cpp [hunk 8]
```diff
@@ -1,98 +1,66 @@
...
-    wl_list_insert(&server->outputs, &output->link);
+  wlr_render_texture_with_matrix(rdata->renderer, texture, matrix, 1);
 
-    wlr_output_layout_add_auto(server->output_layout, wlr_output);
-    wlr_output_create_global(wlr_output);
-  }
+  wlr_surface_send_frame_done(surface, rdata->when);
+}
+
+struct wlr_output_layout *ServerOutput::getLayout() const
```
Comment (Gaspard--) [inline src/ServerOutput.cpp]:
```suggestion
struct wlr_output_layout *ServerOutput::getLayout() const noexcept
```

File: src/View.cpp [hunk 9]
```diff
@@ -1,6 +1,27 @@
...
+  wl_signal_add(&xdg_surface->events.destroy, &destroy);
+  SET_LISTENER(View, ViewListeners, request_move, server->xdgShell->xdg_toplevel_request_move);
+  wl_signal_add(&toplevel->events.request_move, &request_move);
+  SET_LISTENER(View, ViewListeners, request_resize, server->xdgShell->xdg_toplevel_request_resize);
+  wl_signal_add(&toplevel->events.request_resize, &request_resize);
+}
+
 namespace ServerView
 {
   void focus_view(View *view, struct wlr_surface *surface)
```
Comment (Gaspard--) [inline src/View.cpp:27]: Can't comment directly on the line, but the check `view == NULL` should be replaced by `!view`

**Review by Gaspard-- — CHANGES_REQUESTED** [review]

**Comment by Gaspard--** [comment]
A lot of warnings about float conversions appear when compiling. I'm ok with taking case of them at a later data.

File: include/Listeners.hpp (hunk 1 above)
Comment (riamon-v) [reply include/Listeners.hpp]: It made more sens to me to isolate the SET_LISTENER macro such as all the listeners that came with it in one dedicated file where you can find again all the listeners you need

File: include/Output.hpp (hunk 2 above)
Comment (riamon-v) [reply include/Output.hpp:27]: Just a simple mistake that went through. See issue #9 that has been made to manage this kind of thing, among other things.

**Review by riamon-v — COMMENTED** [review]
One of the purpose of issue #9

File: include/Output.hpp (hunk 2 above)
Comment (Gaspard--) [reply include/Output.hpp:27]: ok for me.

File: include/Listeners.hpp (hunk 1 above)
Comment (Gaspard--) [reply include/Listeners.hpp]: I feel it makes it harder to know what's inside a class. Also each time you include this file you pull a lot of useless dependencies.

Comment (Gaspard--) [reply include/Listeners.hpp]: (speaking of structs ofc.)

File: src/Keyboard.cpp (hunk 3 above)
Comment (riamon-v) [reply src/Keyboard.cpp:49]: Good question. I don't know if we should unalocated it at the end of the use of Keyboard class or somwhere else tbh

File: include/Listeners.hpp (hunk 1 above)
Comment (riamon-v) [reply include/Listeners.hpp]: You're make a good point here..
So what about the SET_LISTENER macro ?

**Comment by riamon-v** [comment]
> A lot of warnings about float conversions appear when compiling. I'm ok with taking case of them at a later data.

File: src/Keyboard.cpp (hunk 3 above)
Comment (Gaspard--) [reply src/Keyboard.cpp:49]: Opened #12 for this.

File: include/Listeners.hpp (hunk 1 above)
Comment (Gaspard--) [reply include/Listeners.hpp]: It should stay here.

**Review by Gaspard-- — APPROVED** [review]

**Pull request merged**